### PR TITLE
feat(firebase_auth): v1 rework

### PR DESCRIPTION
**Pull request by Salakar**
# Firebase Authentication

Overall, Firebase Auth has been heavily reworked to bring it inline with the federated plugin setup along with adding new features, documentation and many more unit and end-to-end tests. The API has mainly been kept the same, however there are some breaking changes.

### General

- **BREAKING**: The `FirebaseUser` class has been renamed to `User`.
- **BREAKING**: The `AuthResult` class has been renamed to `UserCredential`.
- **NEW**: The `ActionCodeSettings` class is now consumable on all supporting methods.
  - **NEW**: Added support for the `dynamicLinkDomain` property.
- **NEW**: Added a new `FirebaseAuthException` class (extends `FirebaseException`).
  - All errors are now returned as a `FirebaseAuthException`, allowing you to access the code & message associated with the error. 
  - In addition, it is now possible to access the `email` and `credential` properties on exceptions if they exist.

### `FirebaseAuth`

- **BREAKING**: Accessing the current user via `currentUser()` is now synchronous via the `currentUser` getter.
- **BREAKING**: `isSignInWithEmailLink()` is now synchronous.
- **DEPRECATED**: `FirebaseAuth.fromApp()` is now deprecated in favor of `FirebaseAuth.instanceFor()`.
- **DEPRECATED**: `onAuthStateChanged` has been deprecated in favor of `authStateChanges()`.
- **NEW**: Added support for `idTokenChanges()` stream listener.
- **NEW**: Added support for `userChanges()` stream listener.
  - The purpose of this API is to allow users to subscribe to all user events without having to manually hydrate app state in cases where a manual reload was required (e.g. `updateProfile()`). 
- **NEW**: Added support for `applyActionCode()`.
- **NEW**: Added support for `checkActionCode()`.
- **NEW**: Added support for `verifyPasswordResetCode()`.
- **NEW**: Added support for accessing the current language code via the `languageCode` getter.
- **NEW**: `setLanguageCode()` now supports providing a `null` value.
  - On web platforms, if `null` is provided the Firebase projects default language will be set.
  - On native platforms, if `null` is provided the device language will be used.
- **NEW**: `verifyPhoneNumber()` exposes a `autoRetrievedSmsCodeForTesting` property.
  - This allows developers to test automatic SMS code resolution on Android devices during development.
- **NEW** (iOS): `appVerificationDisabledForTesting`  setting can now be set for iOS.
  - This allows developers to skip ReCaptcha verification when testing phone authentication.
- **NEW** (iOS): `userAccessGroup` setting can now be set for iOS & MacOS.
  - This allows developers to share authentication states across multiple apps or extensions on iOS & MacOS. For more information see the [Firebase iOS SDK documentation](https://firebase.google.com/docs/auth/ios/single-sign-on).

### `User`

- **BREAKING**: Removed the `UpdateUserInfo` class when using `updateProfile` in favor of named arguments.
- **NEW**: Added support for `getIdTokenResult()`.
- **NEW**: Added support for `verifyBeforeUpdateEmail()`.
- **FIX**: Fixed several iOS crashes when the Firebase SDK returned `nil` property values.
- **FIX**: Fixed an issue on Web & iOS where a users email address would still show after unlinking the email/password provider.

### `UserCredential`

- **NEW**: Added support for accessing the users `AuthCredential` via the `credential` property.

### `AuthProvider` & `AuthCredential`

- **DEPRECATED**: All sub-class (e.g. `GoogleAuthProvider`) `getCredential()` methods have been deprecated in favor of `credential()`.
  - **DEPRECATED**:  `EmailAuthProvider.getCredentialWithLink()` has been deprecated in favor of `EmailAuthProvider.credentialWithLink()`.
- **NEW**: Supporting providers can now assign scope and custom request parameters.
  - The scope and parameters will be used on web platforms when triggering a redirect or popup via `signInWithPopup()` or `signInWithRedirect()`.

Closes

- [x] https://github.com/FirebaseExtended/flutterfire/issues/1456
- [x] https://github.com/FirebaseExtended/flutterfire/issues/217
- [x] https://github.com/FirebaseExtended/flutterfire/issues/1949
- [x] https://github.com/FirebaseExtended/flutterfire/issues/2182
- [x] https://github.com/FirebaseExtended/flutterfire/issues/2182
- [x] https://github.com/FirebaseExtended/flutterfire/issues/3095
- [x] https://github.com/FirebaseExtended/flutterfire/issues/2383
- [x] https://github.com/FirebaseExtended/flutterfire/issues/48
- [x] https://github.com/FirebaseExtended/flutterfire/issues/1537
- [x] https://github.com/FirebaseExtended/flutterfire/issues/1456
- [x] https://github.com/FirebaseExtended/flutterfire/issues/1318
- [x] https://github.com/FirebaseExtended/flutterfire/issues/1934
- [x] https://github.com/FirebaseExtended/flutterfire/issues/1767
 - Notes: Working as intended
- [x] https://github.com/FirebaseExtended/flutterfire/issues/221
- [x] https://github.com/FirebaseExtended/flutterfire/issues/2319
- [x] https://github.com/FirebaseExtended/flutterfire/issues/2866

### CODE REVIEW

File: packages/firebase_auth/firebase_auth/lib/src/firebase_auth.dart [hunk 1]
```diff
@@ -74,269 +175,424 @@ class FirebaseAuth {
...
+      _pipeStreamChanges(_delegate.idTokenChanges());
+
+  /// Notifies about changes to any user updates.
+  ///
+  /// This is a superset of both [authStateChanges] and [idTokenChanges]. It provides
+  /// events on all user changes, such as when credentials are linked, unlinked and
+  /// when updates to the user profile are made. The purpose of this Stream is to
+  /// for listening to realtime updates to the user without manually having to call
+  /// [reload] and then rehydrating changes to your application.
+  Stream<User> userChanges() => _pipeStreamChanges(_delegate.userChanges());
```
Comment (kroikie) [inline packages/firebase_auth/firebase_auth/lib/src/firebase_auth.dart:240]: Nit: Is there anyway to have this extracted to another part of the SDK so that it is clear that this is not a Firebase method?

File: packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java [hunk 2]
```diff
@@ -123,770 +143,1119 @@ public void onDetachedFromActivity() {
...
-            Map<String, Object> arguments = new HashMap<>();
-            arguments.put("handle", handle);
-            arguments.put("verificationId", verificationId);
-            arguments.put("forceResendingToken", forceResendingToken.hashCode());
-            channel.invokeMethod("phoneCodeSent", arguments);
-          }
+    // If the credential map contains a token, it means a native one has been stored
+    if (credentialMap.get("token") != null) {
+      int token = (int) credentialMap.get("token");
+      AuthCredential credential = mAuthCredentials.get(token);
```
Comment (kroikie) [inline packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java:221]: nit: We can do this in another PR but would like us to move away from the `m` prefix on the Android side.

File: packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java [hunk 3]
```diff
@@ -123,770 +143,1119 @@ public void onDetachedFromActivity() {
...
 
-    result.success(null);
+    String providerId = (String) credentialMap.get("providerId");
+    String secret = (String) credentialMap.get("secret");
+    String idToken = (String) credentialMap.get("idToken");
+    String accessToken = (String) credentialMap.get("accessToken");
+    String rawNonce = (String) credentialMap.get("rawNonce");
+
+    switch (providerId) {
+      case "password":
```
Comment (kroikie) [inline packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java:237]: nit: Should we be using constants here? eg: PROVIDER_PASSWORD

File: packages/firebase_auth/firebase_auth/example/test_driver/instance_e2e.dart [hunk 4]
```diff
@@ -0,0 +1,740 @@
...
+        await ensureSignedOut();
+
+        Stream<User> stream = auth.authStateChanges();
+        Stream<User> stream2 = auth.authStateChanges();
+
+        StreamSubscription subscription =
+            stream.listen(expectAsync1((User user) {}, count: 2));
+
+        StreamSubscription subscription2 =
+            stream2.listen(expectAsync1((User user) {}, count: 3));
```
Comment (kroikie) [inline packages/firebase_auth/firebase_auth/example/test_driver/instance_e2e.dart:88]: Question: does this fail if stream2 is only triggered 2 times?

File: packages/firebase_auth/firebase_auth/lib/src/firebase_auth.dart [hunk 5]
```diff
@@ -5,65 +5,166 @@
...
+  ///
+  /// A [FirebaseAuthException] maybe thrown with the following error code:
+  /// - **expired-action-code**:
+  ///  - Thrown if the action code has expired.
+  /// - **invalid-action-code**:
+  ///  - Thrown if the action code is invalid. This can happen if the code is malformed or has already been used.
+  /// - **user-disabled**:
+  ///  - Thrown if the user corresponding to the given action code has been disabled.
+  /// - **user-not-found**:
+  ///  - Thrown if there is no user corresponding to the action code. This may have happened if the user was deleted between when the action code was issued and when this method was called.
```
Comment (kroikie) [inline packages/firebase_auth/firebase_auth/lib/src/firebase_auth.dart:106]: nit: Line length

File: packages/firebase_auth/firebase_auth/lib/src/user.dart [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+  /// - **invalid-verification-id**:
+  ///  - Thrown if the credential is a [PhoneAuthProvider.credential] and the verification ID of the credential is not valid.
+  Future<UserCredential> reauthenticateWithCredential(
+      AuthCredential credential) async {
+    assert(credential != null);
+    return UserCredential._(
+        _auth, await _delegate.reauthenticateWithCredential(credential));
+  }
+
+  // reauthenticateWithPhoneNumber
```
Comment (kroikie) [inline packages/firebase_auth/firebase_auth/lib/src/user.dart:185]: Nit: is this comment expected to be like the others?

File: packages/firebase_auth/firebase_auth/pubspec.yaml [hunk 7]
```diff
@@ -3,7 +3,7 @@ description: Flutter plugin for Firebase Auth, enabling Android and iOS
   authentication using passwords, phone numbers and identity providers
   like Google, Facebook and Twitter.
 homepage: https://github.com/FirebaseExtended/flutterfire/tree/master/packages/firebase_auth/firebase_auth
-version: 0.16.1+2
+version: 0.16.1
```
Comment (kroikie) [inline packages/firebase_auth/firebase_auth/pubspec.yaml:6]: Would this be pre-release initially `0.17.0-dev.1` or similar?

File: packages/firebase_auth/firebase_auth/lib/src/firebase_auth.dart (hunk 1 above)
Comment (Ehesp) [reply packages/firebase_auth/firebase_auth/lib/src/firebase_auth.dart:240]: Where were you thinking? It's an app specific method so needs to be on the instance of the class (where it currently is). Possible to move it, just not sure where.

File: packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java (hunk 2 above)
Comment (Ehesp) [reply packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java:217]: Sure, more of a habit! Will ping you.

File: packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java [hunk 8]
```diff
@@ -123,770 +143,1119 @@ public void onDetachedFromActivity() {
...
+  //  private Map<String, Object> parseMultiFactorInfo(@NonNull MultiFactorInfo multiFactorInfo) {
+  //    Map<String, Object> output = new HashMap<>();
+  //
+  //    output.put("displayName", multiFactorInfo.getDisplayName());
+  //    output.put("enrollmentTimestamp", multiFactorInfo.getEnrollmentTimestamp() * 1000);
+  //    output.put("factorId", multiFactorInfo.getFactorId());
+  //    output.put("uid", multiFactorInfo.getUid());
+  //
+  //    return output;
+  //  }
```
[the remark this answers is not in the document]
Comment (Ehesp) [reply packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java:307]: This is the multi-factor authentication work which I started but left out until iOS & web support it. I can remove if you'd like and add back in at a later date.

File: packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java [hunk 9]
```diff
@@ -123,770 +143,1119 @@ public void onDetachedFromActivity() {
...
+                .setAutoRetrievedSmsCodeForPhoneNumber(phoneNumber, autoRetrievedSmsCodeForTesting);
+          }
+
+          PhoneAuthOptions.Builder phoneAuthOptionsBuilder =
+              new PhoneAuthOptions.Builder(firebaseAuth);
+          phoneAuthOptionsBuilder.setActivity(getActivity());
+          phoneAuthOptionsBuilder.setPhoneNumber(phoneNumber);
+          phoneAuthOptionsBuilder.setCallbacks(callbacks);
+          phoneAuthOptionsBuilder.setTimeout((long) timeout, TimeUnit.MILLISECONDS);
+          //          phoneAuthOptionsBuilder.requireSmsValidation(requireSmsValidation);
```
[the remark this answers is not in the document]
Comment (Ehesp) [reply packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java:810]: Another MFA only property, I'll remove for now.

File: packages/firebase_auth/firebase_auth/example/test_driver/instance_e2e.dart (hunk 4 above)
Comment (Ehesp) [reply packages/firebase_auth/firebase_auth/example/test_driver/instance_e2e.dart:88]: Yeah, it would timeout if the 3rd one never gets called. The `expectAsync` api blocks until it's satisfied or times out.

**Comment by vytautas-pranskunas-** [comment]
Is there a plan when this rework will be released?

**Comment by Ehesp** [comment]
@vytautas-pranskunas- Soon, we're just going through internal review along, finishing up testing/CI and a few iOS changes are needed. Hopefully in the next week or so.

**Comment by vytautas-pranskunas-** [comment]
@Ehesp will onIdTokenChanged solve the problem with token refresh?
Can you read my comment here:
https://github.com/FirebaseExtended/flutterfire/issues/2319
and here
https://stackoverflow.com/questions/63181424/request-retry-best-practices/63181505?noredirect=1#comment111726543_63181505

**Comment by Ehesp** [comment]
Yep, along with `userChanges` too :)

**Comment by kroikie** [comment]
fixes: https://github.com/FirebaseExtended/flutterfire/issues/3095
